dashboard/test/controllers/api/v1/sections_controller_test.rb

```diff
@@ -95,4 +95,39 @@ class Api::V1::SectionsControllerTest < ActionController::TestCase
 
     assert_equal @course.id, json['course_id']
   end
+
+  test "join with invalid section code" do
+    assert_raises(ActiveRecord::RecordNotFound) do
+      post :join, params: {id: 'xxxxxx'}
+    end
+  end
+
+  test "join with nobody signed in" do
+    post :join, params: {id: @word_section.code}
+    assert_response 404
+  end
+
+  test "join with valid section code" do
+    student = create :student
+    sign_in student
+    post :join, params: {id: @word_section.code}
+    assert_response :success
+  end
+
+  test "leave with invalid section code" do
+    assert_raises(ActiveRecord::RecordNotFound) do
+      post :leave, params: {id: 'xxxxxx'}
+    end
+  end
+
+  test "leave with nobody signed in" do
+    post :leave, params: {id: @word_section.code}
+    assert_response 404
+  end
+
+  test "leave with valid section code" do
+    sign_in @word_user_1
+    post :leave, params: {id: @word_section.code}
+    assert_response :success
+  end
```

Comment:
Good ideas.

Joining a section that you're already a member of just works (https://github.com/code-dot-org/code-dot-org/blob/b8bf1426f90ed62a3e06124393db3d8e6baf28f2/dashboard/app/models/sections/section.rb#L135-L138) but I've added a test for leaving a section you're not a member of.

The current server behaviour which I copied didn't involve checking for a teacher-managed account, so at the moment it's just enforced client-side through hiding some UI.  But happy to discuss this as a future improvement.